roofit/xroofit/src/xRooNode.cxx

```diff
@@ -5985,120 +7355,235 @@ TH1 *xRooNode::BuildHistogram(RooAbsLValue *v, bool empty, bool errors, int binS
+           // technically should check parent is a RooAddPdf, because if was inside a RooRealSumPdf then would be evaluated as a function!
+       }
+
+      // check if we need to do any projecting of other observables
+      RooAbsReal *oldrar = nullptr;
+      auto _obs = obs();
+
+      for (auto o : _obs) {
+         if (auto rr = o->get<RooRealVar>(); rr && rr->hasRange("coordRange")) {
+            rr->removeRange("coordRange");
+         }
+      }
+      coords(); // loads current coordinates and populates coordRange, if any
+
+      if (auto a = dynamic_cast<RooAbsArg *>(v))
+         _obs.get<RooArgList>()->remove(*a);
+      if (!_obs.get<RooArgList>()->empty()) {
+         oldrar = rar;
+         normSet.add(*_obs.get<RooArgList>());
+         // check if any obs are restricted range
+         bool hasRange = false;
+         for (auto o : normSet) {
+            if (auto rr = dynamic_cast<RooRealVar *>(o); rr && rr->hasRange("coordRange")) {
+               hasRange = true;
+               break;
+            }
+         }
+         if (p) {
+            // need to handle special case of RooSimultaneous ... each pdf needs individually projecting over just its dependent obs
+            if(auto s = dynamic_cast<RooSimultaneous*>(p)) {
+               auto newrar = new RooSimultaneous("projSim","projSim",const_cast<RooAbsCategoryLValue&>(s->indexCat()));
+               for(auto pdf : bins()) {
+                  //auto _pdf = pdf->get<RooAbsPdf>()->createProjection(*pdf->get<RooAbsPdf>()->getObservables(*_obs.get<RooArgList>()));
+                  auto _pdf = new xRooProjectedPdf(TString::Format("%s_projection",pdf->GetName()),"",*pdf->get<RooAbsPdf>(),*pdf->get<RooAbsPdf>()->getObservables(*_obs.get<RooArgList>()));
+                  if (hasRange) {
+                     dynamic_cast<RooAbsPdf *>(_pdf)->setNormRange("coordRange");
+                  }
+                  newrar->addPdf(*_pdf,pdf->coords()["channelCat"]->get<RooCategory>()->getLabel());
+               }
+               rar = newrar;
+            } else {
+               rar = p->createProjection(*_obs.get<RooArgList>()); // TODO should use xRooProjectedPdf here too, because not fixed range and extend behaviour of RooProjectedPdf in ROOT yet
+               if (hasRange) {
+                  dynamic_cast<RooAbsPdf *>(rar)->setNormRange("coordRange");
+               }
+            }
+            if(hasRange) p->setNormRange("coordRange"); // should get cleared when we sterilize
+         } else {
+            if (hasRange) {
+                // commented out passing of normset so that getVal of non-pdf is always a 'raw' value (needed for raw eval of RooRealSumPdf)
+               rar =
+                  rar->createIntegral(*_obs.get<RooArgList>(), /*RooFit::NormSet(normSet),*/ RooFit::Range("coordRange"));
+            } else {
+               rar = rar->createIntegral(*_obs.get<RooArgList>()/*, RooFit::NormSet(normSet)*/);
```

Comment:
This doesn't build in memory safe mode, because `createIntegral()` returns an owning pointer to a RooAbsReal.

Who is owning that RooAbsReal here? Is it an actual memory leak?

Reply:
this bit of code is a bit convoluted I will admit and needs to be tidied up at some point (it got in this state because the method has evolved over time ... this recent stuff is to support projecting PDFs out into 1d histograms). I dont think its a memory leak as things should be cleaned up later but at some point yes I should move this over to smart pointers. I've done a unique_ptr and a `release()` call for now though

Reply:
Okay, that's a good solution for now, and thanks for giving some context!